pyclam/criterion.py

```diff
@@ -130,38 +130,89 @@ def __call__(self, root: Cluster) -> Set[Cluster]:
+    clusters: List[Cluster] = [child for child in manifold.root.children]
+    while clusters:
+        new_clusters: List[Cluster] = list()
+        for cluster in clusters:
+            if cluster.children:
+                subtree_auc = [predicted_auc[c] for c in cluster.descendents]
+                if predicted_auc[cluster] >= np.percentile(subtree_auc, q=75):
+                    graph.add(cluster)
+                else:
+                    new_clusters.extend(cluster.children)
+            else:
+                graph.add(cluster)
+        clusters = new_clusters
+    return graph
+
+
+def _ranked_selection(manifold: Manifold, predicted_auc: Dict[Cluster, float]) -> Set[Cluster]:
+    """ Rank-orders clusters by predicted_auc and peel off the top to build the graph.
+    """
+    rankings: Dict[Cluster, float] = {cluster: 0 for cluster in predicted_auc.keys()}
+    for leaf in manifold.layers[-1].clusters:
+        # sort ancestors by predicted auc in non-increasing order, break ties by smaller depth
+        ancestors = list(sorted(manifold.ancestry(leaf), key=lambda c: (predicted_auc[c], c), reverse=True))
+        # add index of cluster in sorted list of ancestors to rankings
+        for i, cluster in enumerate(ancestors):
+            rankings[cluster] += (i * cluster.cardinality)
+    rankings.pop(manifold.select(''))  # remove root from rankings because ratios are undefined
+
+    # TODO: misses some points/clusters. breaks graph invariant
+    # sort clusters by mean ranking in non-decreasing order
+    potentials: List[Cluster] = list(sorted(list(rankings.keys()), key=lambda c: (rankings[c] / c.cardinality, c)))
+    selected: Set[Cluster] = set()
+    excluded: Set[Cluster] = set()
+    # select best clusters for graph
+    for cluster in potentials:
+        if cluster in excluded:
+            continue
+        else:
+            selected.add(cluster)
+            excluded.add(cluster)
+            excluded.update(cluster.ancestors)
+            excluded.update(cluster.descendents)
+
+    # TODO: remove this after fixing the missed clusters in the initial pass
+    missed: Set[int] = set(manifold.root.argpoints) - {p for c in selected for p in c.argpoints}
+    for layer in manifold.layers[1:]:
+        for cluster in layer.clusters:
+            if all((p in missed for p in cluster.argpoints)):
+                selected.add(cluster)
+                missed.difference_update(cluster.argpoints)
+
+    return selected
+
+
+SELECTION_MODES = {
+    'percentile': _percentile_selection,
+    'ranked': _ranked_selection,
+}
+
+
 class LinearRegressionConstants(SelectionCriterion):
     """ Uses constants from a meta-ml model using linear regression.
     There are 6 constants, one for each of the ratios and ema_ratios in Cluster.
     """
 
-    def __init__(self, constants: Union[np.array, List[float]]):
-        num_constants = 6
+    def __init__(self, constants: Union[np.array, List[float]], *, mode: str = 'ranked'):
```

Comment:
docstring should specify what modes are allowed

Reply:
fixed